torch/nn/parallel/distributed.py

```diff
@@ -222,21 +222,27 @@ def __init__(self, module, device_ids=None,
         )
 
         self.is_multi_device_module = len({p.device for p in module.parameters()}) > 1
-        self.is_cuda = all([p.device.type == 'cuda' for p in module.parameters()])
-
-        if not self.is_cuda or self.is_multi_device_module:
+        device_type = {p.device.type for p in module.parameters()}
+        assert len(device_type) == 1, (
+            "DistributedDataParallel with multi-device module only works "
+            "with GPU devices, but module parameters locate in {}."
+        ).format(device_type)
+        self.device_type = list(device_type)[0]
+        self.is_cuda = True if self.device_type == 'cuda' else False
```

Comment:
Since this PR is to decouple DDP from CUDA, should we avoid using `is_cuda` instead? Currently, it is also used in `_passing_sync_batchnorm_handle`.

Reply:
This is for backward compatibility. 
I am very glad to remove it. I will check the usage of the `is_cuda`.
Do you have any other concern about removing the `is_cuda` other than `_passing_sync_batchnorm_handle`